package-best-practices.md

```diff
@@ -0,0 +1,46 @@
+
+Although the dbt Labs team completes a cursory review of new packages before adding them to the Package Hub, ultimately you are responsible for your package's performance.
+
+To avoid ambiguity, the bullets below are pretty formal, but we are a friendly bunch! If you need help preparing a package for deployment to the dbt Package Hub, reach out in [#package-ecosystem](https://getdbt.slack.com/archives/CU4MRJ7QB/) in the [dbt Community Slack](https://getdbt.com/community).
+
+---
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+### First run experience
+- Packages SHOULD contain a README file which explains how to get started with the package and customise its behaviour.
+- The documentation SHOULD indicate which data warehouses are expected to work with this package.
+
+### Customisability 
+- Packages MUST NOT hard-code table references, and MUST use `ref` or `source` instead.
+- Packages MAY provide a mechanism (such as variables) to enable/disable certain subsets of functionality.
+- Packages for data transformation:
+    - SHOULD provide a mechanism (such as variables) to customise the location of source tables.
+    - SHOULD NOT assume database/schema names in sources. 
+    - MAY assume table names, particularly if the package was built to support tables created by a known tool.
+
+### Dependencies
+#### Dependencies on dbt Core
+- Packages requiring a minimum version of `dbt-core` MUST declare it in the `require-dbt-version` property of `dbt_project.yml`.
+- Packages requiring a minimum version of `dbt-core` SHOULD allow all subsequent minor and patch releases of that major version. 
+    - For example, a package which depends on functionality added in dbt Core 1.2 SHOULD set its `require-dbt-version` property to `[">=1.2.0", "<2.0.0"]`.
+#### Dependencies on other packages defined in `packages.yml`:
+- Packages SHOULD import their dependencies from the dbt Package Hub when available, as opposed to a `git` installation.
+- Packages SHOULD NOT pin to a patch version of their imported package unless they are aware of an incompatibility.
+- Packages importing another package which has reached major version one or later SHOULD allow all subsequent minor and patch releases of that major version. 
+- Packages importing another package which is major version zero MAY pin to the current minor version only.
+
+### Interoperability
+- Packages MUST NOT override dbt Core behaviour in such a way as to impact other dbt resources (models, tests, etc) not provided by the package.
+- Packages MUST use the cross-database macros built into dbt Core where available, such as `{{ datediff }}` and `{{ type_string() }}`.
```

Comment:
`MUST` or `SHOULD`? Could a BigQuery-specific spend package BigQuery-isms? Or what if the current cross-database implementation doesn't cover some edge case that the package needs or wants to accommodate?

Optional ideas for re-wording slightly:
```suggestion
- Packages SHOULD use the [cross-database macros](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros) built into dbt Core where available, such as `{{ except() }}` and `{{ type_string() }}`.
```

Reply:
Went back and forth on this and maybe came down too hard 🤔. How about packages intended for use with multiple warehouses MUST...

I like the link to the docs

Reply:
If there's an edge case that the macros don't cover, then I think the "where available" works as an escape hatch

Reply:
Love your suggestion to introduce the nuanced language of [RFC 6919](https://datatracker.ietf.org/doc/html/rfc6919) where it is useful! 🤩 

News flash: I'm not an expert on the interpretation and application of RFC 2119 or 6919. The following is all informed by my nascent understanding and interpretations.

After reading the ["Guidance in the use of these Imperatives"](https://datatracker.ietf.org/doc/html/rfc2119#section-6) section of RFC 2119, it feels like `SHOULD` is the applicable imperative here. The following stood out to me in particular:
> they must not be used to try to impose a particular method on implementors where the method is not required for interoperability.

A package could ignore some of our "musts" and operate just fine. I didn't check, but I'm guessing there are examples currently in the [Hub](https://hub.getdbt.com/) that would back this up.

But most (not all) of our current usage of `MUST` seems like it should be `SHOULD` instead. `SHOULD` appears to be a **really** strong word in the spec! It shouldn't be confused at all with `OPTIONAL`.

So I think we shouldn't feel like we are pandering or compromising on our expectations in any way when we use `SHOULD`. We should feel like we're laying down something strong, confident, and very assertive.

Reply:
Tragically, 6919 was published on **1 April** 2013 and was never ratified. (See also https://datatracker.ietf.org/doc/html/rfc2324)

Happy to roll with SHOULD here